src/IssueViz.Security/Taint/ServerSentEvents/TaintServerEventsListener.cs

```diff
@@ -0,0 +1,115 @@
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+
+using System.Threading.Tasks;
+using SonarLint.VisualStudio.Core;
+using System.Threading;
+using System.ComponentModel.Composition;
+using SonarLint.VisualStudio.Integration;
+using SonarQube.Client.Models.ServerSentEvents.ClientContract;
+using System;
+
+namespace SonarLint.VisualStudio.IssueVisualization.Security.Taint.ServerSentEvents
+{
+    /// <summary>
+    /// Consumes <see cref="ITaintServerEventSource"/> and handles <see cref="ITaintServerEvent"/> coming from the server.
+    /// </summary>
+    internal interface ITaintServerEventsListener : IDisposable
+    {
+        Task ListenAsync();
+    }
+
+    [Export(typeof(ITaintServerEventsListener))]
+    [PartCreationPolicy(CreationPolicy.Shared)]
+    internal sealed class TaintServerEventsListener : ITaintServerEventsListener
+    {
+        private readonly IStatefulServerBranchProvider serverBranchProvider;
+        private readonly ITaintServerEventSource taintServerEventSource;
+        private readonly ITaintStore taintStore;
+        private readonly IThreadHandling threadHandling;
+        private readonly ITaintIssueToIssueVisualizationConverter taintToIssueVizConverter;
+        private readonly ILogger logger;
+        private readonly CancellationTokenSource cancellationTokenSource;
+
+        [ImportingConstructor]
+        public TaintServerEventsListener(
+            IStatefulServerBranchProvider serverBranchProvider,
+            ITaintServerEventSource taintServerEventSource,
+            ITaintStore taintStore,
+            IThreadHandling threadHandling,
+            ITaintIssueToIssueVisualizationConverter taintToIssueVizConverter,
+            ILogger logger)
+        {
+            this.serverBranchProvider = serverBranchProvider;
+            this.taintServerEventSource = taintServerEventSource;
+            this.taintStore = taintStore;
+            this.threadHandling = threadHandling;
+            this.taintToIssueVizConverter = taintToIssueVizConverter;
+            this.logger = logger;
+
+            cancellationTokenSource = new CancellationTokenSource();
+        }
+
+        public async Task ListenAsync()
+        {
+            await threadHandling.SwitchToBackgroundThread();
+
+            while (!cancellationTokenSource.IsCancellationRequested)
+            {
+                var taintServerEvent = await taintServerEventSource.GetNextEventOrNullAsync();
+
+                switch (taintServerEvent)
+                {
+                    case null:
+                    {
+                        return;
```

Comment:
Can we rely on it being null when and only when we're done?

Reply:
That should be the behavior of IServerSentEventSource - it's supposed to block indefinitely until Writer.Complete is called on the channel, which is done only on dispose.